### ISSUE TITLE
Emulating the iPhone X

### ISSUE THREAD

**Issue by umpox** [ChromeDevTools/devtools-frontend#34]
I see there are device frames for a variety of devices. https://github.com/ChromeDevTools/devtools-frontend/tree/master/front_end/emulated_devices

Do you think there's a possibility of adding the iPhone X frame in the future? As it has an usual shape I feel it would be a useful addition.

**Comment by jonathanarbely** [comment]
These devices are now part of Chrome's emulator. See [this commit](https://github.com/ChromeDevTools/devtools-frontend/commit/5c2aec69661ae3aadaf382e35ec321ce674bf5bd).

**Comment by JPT62089** [comment]
Agreed, the unique special fruitcake shape needs to be added to the Device Emulation. Notch and speaker included.

**Comment by sskogen** [comment]
@jonathanarbely The device is part of the commit, but not the device frame

**Comment by gyasikn** [comment]
I can see all the frames for other devices but iphone x does not have a frame. Anyone know how to fix this or is it just not available?

**Comment by paulirish** [comment]
cc @psybuzz for visibility.


the new.crbug.com bug tracker is the right place for this request. sorry you ended up here.

however technically we can add the device frame, but likely will not be able to emulate the notch behavior very well. our screen emulation is definitely not designed for that.

**Comment by sskogen** [comment]
@paulirish A proper emulation of the notch behavior is not critical for me personally, but I can't answer for all here. Would love to just see the frame and maybe capture screenshots including the frame, like for the other devices.

**Comment by israelroldan** [comment]
@gyasikn based on the commit shared by @jonathanarbely it appears that it is not available. Most likely due to it is not as easy as just adding the "outline" property, but rather as @sskogen mentions, the notch behavior is key to properly emulate it.

**Comment by blogcraft** [comment]
I hope a notch support is considered someday, because a lot of new devices are coming with it and also Android P will have native support for at least 3 different notch sizes.

**Comment by adampfoster** [comment]
You can vote / star for this bug, so it will receive attention https://bugs.chromium.org/p/chromium/issues/detail?id=838829

**Comment by gdarcan** [comment]
iPhoneX();
window.onresize = window.onorientationchange = function() {
	//Your other functions
	setTimeout(iPhoneX,100);
}

function iPhoneX() {
	if (window.innerWidth == 375 && window.innerHeight == 812) {
		if (!document.getElementById('iphone_layout')) {
			var img = document.createElement('img');
			img.id = 'iphone_layout';
			img.style.position = 'fixed';
			img.style.zIndex = 9999;
			img.style.pointerEvents = 'none';
			img.src = 'img/iphonex.png'
			document.body.insertBefore(img,document.body.children[0]);
		}
	} else if (document.getElementById('iphone_layout')) {
		document.body.removeChild(document.getElementById('iphone_layout'));
	}
}

![iphonex](https://user-images.githubusercontent.com/19612632/41588096-1ac4c128-73b1-11e8-9fc7-db9ab61903ed.png)
![screenshot](https://user-images.githubusercontent.com/19612632/41588276-9a64f7b8-73b1-11e8-8add-7398775697c7.jpg)

**Comment by branditodesign** [comment]
I've only been able to see about 4-5 device frames for as long as I can remember, am I missing something?

**Comment by chasebank** [comment]
This is an old issue but seems like it should be a bit higher priority now that the notch has fully taken hold in most of the flagship devices, including Google's own Pixel 3.

Is the scope of this issue limited to just the display of notched frames? Or the general support of the notch?

The previously linked Chromium issue reads like it's just concerned with the frames.

But personally I'm far less concerned with the frame (which appears to simply be a decorative overlay) and more concerned with whether or not the actual notch is being observed by the browser. As a user, I have no real way of knowing.

The frame overlay is a useful visual, but if the dev tool isn't actually recognizing a notch, then it's rather useless for testing layout IMO.

In testing [environment variables](https://webkit.org/blog/7929/designing-websites-for-iphone-x/) there doesn't seem to be any effect in Chrome 69.0

Is the safe area communicated through the user agent? If so, I would _think_ this is something the browser tools can emulate.

**Comment by gdarcan** [comment]
> This is an old issue but seems like it should be a bit higher priority now that the notch has fully taken hold in most of the flagship devices, including Google's own Pixel 3.
> 
> Is the scope of this issue limited to just the display of notched frames? Or the general support of the notch?
> 
> The previously linked Chromium issue reads like it's just concerned with the frames.
> 
> But personally I'm far less concerned with the frame (which appears to simply be a decorative overlay) and more concerned with whether or not the actual notch is being observed by the browser. As a user, I have no real way of knowing.
> 
> The frame overlay is a useful visual, but if the dev tool isn't actually recognizing a notch, then it's rather useless for testing layout IMO.
> 
> In testing [environment variables](https://webkit.org/blog/7929/designing-websites-for-iphone-x/) there doesn't seem to be any effect in Chrome 69.0
> 
> Is the safe area communicated through the user agent? If so, I would _think_ this is something the browser tools can emulate.

@chasebank , unfortunately, browser doesn't recognize if there's a notch or not (afaik). If they would ask me, I would propose an idea that there should be an extenstion something like window.notch and first of all, it should provide true|false if there's a notch. Also maybe in accordance with phone model, color of the notch (white|black maybe best matching HEX code), and finally (and absolutely it should provide) path of the notch  smtg like [[x1,y1],[x2,y2], ... ]. But I think this is a far away dream and I don't expect anything like that soon.

**Comment by ElectricCookie** [comment]
Going a step further than @gdarcan 's solution. Simply paste this code:
```
iPhoneX();
['resize','orientationchange'].forEach(function(evt) {
    window.addEventListener(evt,iPhoneX,{passive:true});
});

function iPhoneX() {
	if (window.innerWidth == 375 && window.innerHeight == 812) {
		if (!document.getElementById('iphone_layout')) {
			var img = document.createElement('img');
			img.id = 'iphone_layout';
			img.style.position = 'fixed';
			img.style.zIndex = 9999;
			img.style.pointerEvents = 'none';
			img.src = 'https://user-images.githubusercontent.com/19612632/41588096-1ac4c128-73b1-11e8-9fc7-db9ab61903ed.png';
			document.body.insertBefore(img,document.body.children[0]);
		}
	} else if (document.getElementById('iphone_layout')) {
		document.body.removeChild(document.getElementById('iphone_layout'));
	}
}
```

Into [this bookmarklet generator](https://mrcoles.com/bookmarklet/ ). (I was too lazy to make a proper page to serve the bookmarklet.) To get the frame to apply to any page by pressing a single button. 

🥂

**Comment by texano00** [comment]
@ElectricCookie thankss! You've saved me

**Comment by lvl99** [comment]
> It would be great if the iPhone X emulator in Chrome DevTools would support `viewport-fit=cover` / `constant(safe-area-inset-top)` / `env(safe-area-inset-bottom)` attributes.
> BTW, is there any convenient way to debug all these `safe-area` stuff besides the Xcode iPhone X emulator?

This would be great as it would allow our team of PC, Mac and Linux web developers the ability to test the `safe-area-inset-*` variables with our CSS in a consistent environment.
